File: code/Python/Chapter7/7-2/big_int_add.py
def big_int_add(num1, num2):
    # pad the shorter number with leading zeros
    if len(num1) > len(num2):
        num2 = num2.zfill(len(num1))
    else:
        num1 = num1.zfill(len(num2))

    result = ""
    carry = 0
    for i in range(len(num1) - 1, -1, -1):
        digit_sum = int(num1[i]) + int(num2[i]) + carry
        carry = digit_sum // 10
        digit = digit_sum % 10
        result = str(digit) + result
    
    if carry > 0:
        result = str(carry) + result
    return result

print(big_int_add("426709752318", "95481253129"))

File: code/C++/Chapter7/7-2/big_int_add.cpp
#include <iostream>
#include <string>

using namespace std;

string big_int_add(string num1, string num2) {
    // pad the shorter number with leading zeros
    if (num1.size() < num2.size()) {
        num1.insert(0, num2.size() - num1.size(), '0');
    } else {
        num2.insert(0, num1.size() - num2.size(), '0');
    }

    string result;
    int carry = 0;
    for (int i = num1.size() - 1; i >= 0; i--) {
        int digit_sum = (num1[i] - '0') + (num2[i] - '0') + carry;
        carry = digit_sum / 10;
        int digit = digit_sum % 10;
        result.insert(0, 1, digit + '0');
    }

    if (carry > 0) {
        result.insert(0, 1, carry + '0');
    }
    return result;
}

int main() {
    cout << big_int_add("426709752318", "95481253129") << endl;
    return 0;
}